hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestRatisUnderReplicationHandler.java

```diff
@@ -0,0 +1,243 @@
+          DatanodeDetails dn = invocationOnMock.getArgument(0);
+          return new NodeStatus(dn.getPersistedOpState(),
+              HddsProtos.NodeState.HEALTHY);
+        });
+  }
+
+  /**
+   * When the container is under replicated even though there's a pending
+   * add, the handler should create replication commands.
+   */
+  @Test
+  public void testUnderReplicatedWithMissingReplicasAndPendingAdd()
+      throws IOException {
+    Set<ContainerReplica> replicas
+        = createReplicas(container.containerID(), State.CLOSED, 0);
+    List<ContainerReplicaOp> pendingOps = ImmutableList.of(
+        ContainerReplicaOp.create(ContainerReplicaOp.PendingOpType.ADD,
+            MockDatanodeDetails.randomDatanodeDetails(), 0));
+
+    ContainerHealthResult.UnderReplicatedHealthResult healthResult =
+        Mockito.mock(ContainerHealthResult.UnderReplicatedHealthResult.class);
+    Mockito.when(healthResult.getContainerInfo()).thenReturn(container);
+
+    testProcessing(replicas, pendingOps, healthResult, 2, 1);
+  }
+
+  /**
+   * When the container is under replicated and unrecoverable (no replicas
+   * exist), the handler will not create any commands.
+   */
+  @Test
+  public void testUnderReplicatedAndUnrecoverable() throws IOException {
+    ContainerHealthResult.UnderReplicatedHealthResult healthResult =
+        Mockito.mock(ContainerHealthResult.UnderReplicatedHealthResult.class);
+    Mockito.when(healthResult.getContainerInfo()).thenReturn(container);
+
+    testProcessing(Collections.emptySet(), Collections.emptyList(),
+        healthResult, 2, 0);
+  }
+
+  /**
+   * The container is currently under replicated, but there's a pending add
+   * that will make it sufficiently replicated. The handler should not create
+   * any commands.
+   */
+  @Test
+  public void testUnderReplicatedFixedByPendingAdd() throws IOException {
+    Set<ContainerReplica> replicas
+        = createReplicas(container.containerID(), State.CLOSED, 0, 0);
+    List<ContainerReplicaOp> pendingOps = ImmutableList.of(
+        ContainerReplicaOp.create(ContainerReplicaOp.PendingOpType.ADD,
+            MockDatanodeDetails.randomDatanodeDetails(), 0));
+
+    ContainerHealthResult.UnderReplicatedHealthResult healthResult =
+        Mockito.mock(ContainerHealthResult.UnderReplicatedHealthResult.class);
+    Mockito.when(healthResult.getContainerInfo()).thenReturn(container);
+
+    testProcessing(replicas, pendingOps, healthResult, 2, 0);
+  }
+
+  /**
+   * The container is under-replicated because a DN is decommissioning. The
+   * handler should create replication command.
+   */
+  @Test
+  public void testUnderReplicatedBecauseOfDecommissioningReplica()
+      throws IOException {
+    Set<ContainerReplica> replicas = ReplicationTestUtil
+        .createReplicas(Pair.of(DECOMMISSIONING, 0), Pair.of(IN_SERVICE, 0),
+            Pair.of(IN_SERVICE, 0));
+
+    ContainerHealthResult.UnderReplicatedHealthResult healthResult =
```

Comment:
Would be good to create a method to return this mocked HealthResult, but see my earlier comment on the other class - maybe we can make it go away from the interface, as it doesn't see to be too useful in the end.

Reply:
Extracted out a method for this.